feat: asset registry

sander2 opened the pull request:
This is meant to address https://github.com/open-web3-stack/open-runtime-module-library/issues/721.

This PR will need cleanup, tests, etc, but I'm opening this as a draft now to get some feedback regarding the direction. Some things to note about the implementation:

- In the issue I raised two possible implementations: either we can provide some of the metadata fields by default, or we leave it fully customizable. I went with the latter approach in this PR.
- The only dispatchable function is one to register (or update) an asset. 
- The metadata is fully customizable - the only requirement being that there is a way to extract a MultiAsset from it. This is needed to be able to map xcm assets to assets in this pallet
- The assigned asset id is customizable. The last assigned asset id is stored, so you can either derive a sequential id or base it purely on the metadata
- Some traits and trait implementations are provided to allow for easy WeightTrader implementations. Unfortunately because `WeightTrader::refund_weight` allows only a single asset to be refunded, the default implementation does not support paying with different assets. That is, it can pay with any asset in the registry, but only one asset per xcm message. I opened https://github.com/paritytech/polkadot/issues/5364 to see if `refund_weight` can refund multiple assets in the future. 



An example of usage:
``` rust
// the metadata definition...
#[derive(scale_info::TypeInfo, Encode, Decode, Clone, Eq, PartialEq, Debug)]
pub struct AssetMetadata {
    pub location: MultiLocation,
    pub decimals: u32,
    pub fee_per_second: u128,
    pub name: Vec<u8>,
    pub symbol: Vec<u8>,
}
impl Into<MultiLocation> for AssetMetadata {
    fn into(self) -> MultiLocation {
        self.location
    }
}

// the config...
impl orml_asset_registry::Config for Runtime {
    type Event = Event;
    type AssetMetadata = AssetMetadata;
    type AssetId = u32;
    type AuthorityOrigin = EnsureRoot<AccountId>;
    type AssignAsset = SequentialId<u32>;
}

// for paying xcm fees..
pub struct MyFixedConversionRateProvider;
impl FixedConversionRateProvider for MyFixedConversionRateProvider {
    fn get_fee_per_second(location: &MultiLocation) -> Option<u128> {
        let metadata = AssetRegistry::fetch_metadata_by_location(location)?;
        Some(metadata.fee_per_second)
    }
}
pub type Trader = (
    (...)
    AssetRegistryTrader<FixedRateAssetRegistryTrader<MyFixedConversionRateProvider>, ToTreasury>,
);
```

I would love to get some feedback on this. I'm out of office until tuesday though so might not be able to address any comments until then.

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,235 @@
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		pub fn register_asset(origin: OriginFor<T>, metadata: T::AssetMetadata) -> DispatchResult {
```
I think we can add `asset_id: Option<T::AssetId>` parameter so we can specify the asset id if needed.

zqhxuyuan replied:
Agree, parachains like Acala already implements they own asset-registry, If they migrate to orml's asset-registry, seemless migration is helpfull, they could register asset metadata with already exists asset id. In this way, I think `AssetIdAssigner` also needs more trait method for those scene.

sander2 replied:
Done. However, I'm not sure what use case you envision since there is already the `process_asset` trait function that is able to assign an id. Could you clarify?

xlc replied:
Our asset registry supports not only foreign assets, but also native assets. So that asset_id could be an existing one instead of new id.

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,235 @@
+use xcm_executor::{traits::WeightTrader, Assets};
+
+pub trait AssetIdAssigner<AssetId, AssetMetadata> {
+	fn assign_id(asset_metadata: &AssetMetadata, last_id: Option<AssetId>) -> AssetId;
```
make it something like `process_asset(id: Option<AssetId>, metadata: AssetMetadata) -> Result<(AssetId, Metadata), DispatchError>` will be more flexible

not really need to pass last id. the implementation can just read it from this pallet

sander2 replied:
Done. However, without the last id, we can't provide a default `SequentialId` implementation - each chain will need to write an explicit implementation now. Because of that, I would actually be in favor of adding it back in as an arg. What do you think?

xlc replied:
The default SequentialId can still read the last id from the pallet?

xlc commented:
I think we should still require some basic fields like name, symbol, decimals. otherwise the js sdk still need specialized code to handle it for different parachains.

enthusiastmartin commented:
I agree with xlc - that some basic information can be already part of the pallet ( name,symbol, decimals). 

personally, i would not require `Into<MultiLocation>` for asset metadata, and i would not map the location to asset id when asset is registered. 
I'd add ability to set/update/remove location explicitly after asset is already registered.

sander2 commented:
> I think we should still require some basic fields like name, symbol, decimals. otherwise the js sdk still need specialized code to handle it for different parachains.

Sure, I added name, symbol, decimals and multilocation.

> I agree with xlc - that some basic information can be already part of the pallet ( name,symbol, decimals).
> 
> personally, i would not require `Into<MultiLocation>` for asset metadata, and i would not map the location to asset id when asset is registered. I'd add ability to set/update/remove location explicitly after asset is already registered.

When I added default fields, I also added the (multi)location. Are you opposed to this? Do you think it should be an optional item? Or should it be removed as a default field altogether?

enthusiastmartin commented:
> > I think we should still require some basic fields like name, symbol, decimals. otherwise the js sdk still need specialized code to handle it for different parachains.
> 
> Sure, I added name, symbol, decimals and multilocation.
> 
> > I agree with xlc - that some basic information can be already part of the pallet ( name,symbol, decimals).
> > personally, i would not require `Into<MultiLocation>` for asset metadata, and i would not map the location to asset id when asset is registered. I'd add ability to set/update/remove location explicitly after asset is already registered.
> 
> When I added default fields, I also added the (multi)location. Are you opposed to this? Do you think it should be an optional item? Or should it be removed as a default field altogether?

I might be wrong here - but i think, if you set a location for asset and then xcm config is set to map asset and locations - that will make the asset automatically transferable. 

which is probably fine, that's why you add foreign assets in first place.

but not all assets need location, and not all asset are transferable. For example - in HydraDX's xyk pool - we use asset registry to register 'share asset' for each subpool as another asset in the system. However, this asset does not need location nor it can be transferred somewhere else.

sander2 commented:
> I might be wrong here - but i think, if you set a location for asset and then xcm config is set to map asset and locations - that will make the asset automatically transferable.
> 
> which is probably fine, that's why you add foreign assets in first place.
> 
> but not all assets need location, and not all asset are transferable. For example - in HydraDX's xyk pool - we use asset registry to register 'share asset' for each subpool as another asset in the system. However, this asset does not need location nor it can be transferred somewhere else.

Thank you for the clarification, that makes sense indeed. I changed the location to be a `Option<MultiLocation>`, and also added a `set_asset_location` function. In your usecase you'd set the location in the metadata to `None`, while xcm transferable assets can still be registered in a single step using only `register_asset`. The reason why I think it's good to be able to register xcm-transferrable assets in a single call, is that making multiple calls is problematic if you have to go through governance. You could also batch `register_asset` and `set_asset_location`, but the problem there is that the asset id might not be known in advance in all cases

enthusiastmartin commented:
What about existential deposit / min balance ? 

I believe this could be one of the required fields too. 

i think that it would be useful if you could also plug asset registry to eg. orml_tokens's config `ExistentialDeposits`. 
that would mean, it needs `GetByKey` implementation. 

Eg. in our asset-registry [here](https://github.com/galacticcouncil/Basilisk-node/blob/afb76f0c80380fc59d802af62c5a50f85b6ba213/pallets/asset-registry/src/lib.rs#L473).

sander2 commented:
> What about existential deposit / min balance ?
> 
> I believe this could be one of the required fields too.
> 
> i think that it would be useful if you could also plug asset registry to eg. orml_tokens's config `ExistentialDeposits`. that would mean, it needs `GetByKey` implementation.
> 
> Eg. in our asset-registry [here](https://github.com/galacticcouncil/Basilisk-node/blob/afb76f0c80380fc59d802af62c5a50f85b6ba213/pallets/asset-registry/src/lib.rs#L473).

Agreed, added.

sander2 commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,275 @@
+	pub decimals: u32,
+	pub name: Vec<u8>,
+	pub symbol: Vec<u8>,
+	pub location: Option<MultiLocation>,
```
Good question, I'm not sure.. I think `MultiLocation` is easiest to deal with in the parachains, but for off-chain interactions (e.g. reading from js) a versionedLocation might make sense. Anyone else has any thoughts on this?

xlc replied:
Currently `MultiLocation` is referring to the latest version, which will change in future and causing expected issue.
So instead of using the latest, it should use v2 (and move to v3 later if needed) so ensure it doesn't change unexpectedly.

`VersionedMultiLocation` will also avoid this issue here

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,309 @@
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_asset_id)]
+	pub type MultiLocationLookup<T: Config> = StorageMap<_, Twox64Concat, MultiLocation, T::AssetId, OptionQuery>;
```
We can't use xcm::latest::MultiLocation as key format here because it will change.
We can't use VersionedMultiLocation here because it will result inconsistency.
What polkadot xcm pallet doing is using VersionedMultiLocation but always force it to be the latest version and using runtime migration to keep it up to date. But that logic is super complicated.
I will say just use v2::MultiLocation for now and migrate to v3 if needed.

sander2 replied:
Done

sander2 commented:
I cleaned the code up a bit and added some tests, including some that show that it can be used in xcm transfers. From my PoV, this PR is ready to be reviewed.

There is one thing that I still need address: the default weights. For the other pallets, the default weights are generated with `acala/target/release/acala benchmark`. Since this pallet is not included in that repo, should I add benchmark code in the crate itself?

One more note on the `ProcessAsset` trait. By request of @xlc it's returning `(AssetId, Metadata)`, meaning that it can modify the supplied metadata. I'm not certain what the use case of this is, but running the same function again when _modifying_ metadata is problematic, so I only call it in the `register_asset` function and not in `update_asset`.

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::large_enum_variant)]
```
please fix the warning instead of suppress it. in this case you should box VersionedMultiLocation

sander2 replied:
Done, but I think this warning was maybe a false positive, since the other items contain `AssetMetadata`, which has a `Option<VersionedMultiLocation>` field, so as far as I can tell the sizes should have been similar to each other

xlc replied:
It could be false negative. It is known issue that big enum of Call / Event can cause trouble.

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		_phantom: PhantomData<T>,
```
should have a way to register assets in genesis, but it can happen in later PR

sander2 replied:
Hmm yea I actually just ran into a problem when trying to implement that: it requires the `AssetMetadata` to implement Serialize & Deserialize, which `VersionedMultiLocation` does not implement.. We'd have to write a custom implementation, which is certainly possible, but I'd leave it for a future PR indeed

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
```
should use bonded types, but that can happen in later PR

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+
+		/// A filter ran upon metadata registration that assigns an is and
+		/// potentially modifies the supplied metadata.
+		type ProcessAsset: AssetProcessor<Self::AssetId, AssetMetadata<Self::Balance, Self::CustomMetadata>>;
```
maybe call it `OnRegisterAsset` or something similar

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+		metadata: &AssetMetadata<T::Balance, T::CustomMetadata>,
+	) -> DispatchResult {
+		// if the metadata contains a location, set the MultiLocationLookup
+		if let Some(location) = metadata.location.clone() {
```
I think this means when update asset metadata without changing location, the location is still going to be removed & inserted.

Not end of world as we shouldn't update asset metadata every block.

sander2 replied:
You are right. However, since as you say, this shouldn't be called very frequently. We could optimize in the future but for now I prioritized code simplicity over performance

xlc commented on asset-registry/src/impls.rs:
```diff
@@ -0,0 +1,163 @@
+	T: Config<AssetId = AssetId>,
+{
+	fn process_asset(id: Option<AssetId>, asset_metadata: Metadata) -> Result<(AssetId, Metadata), DispatchError> {
+		let next_id = match LastAssetId::<T>::get() {
```
`LastAssetId::<T>::get().unwrap_or_default().saturating_add(One::one())`

xlc reviewed (commented):
not 100% sure if we want all the XCM weight related code in this pallet. I guess we don't really want another crate just for that so it is probably fine.

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+		},
+	}
+
+	/// The total issuance of a token type.
```
bad comment

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+	pub type Metadata<T: Config> =
+		StorageMap<_, Twox64Concat, T::AssetId, AssetMetadata<T::Balance, T::CustomMetadata>, OptionQuery>;
+
+	/// The total issuance of a token type.
```
bad comment

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_asset_id)]
```
the getter name doesn't align with the storage name
also maybe we want to call it `LocationToAssetId`?

xlc commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_metadata)]
```
```suggestion
	#[pallet::getter(fn metadata)]
```

no `get_` in getter
https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

xlc commented:
other than nits it is good for a V1

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+			MultiLocationLookup::<T>::insert(location, asset_id);
+		}
+
+		Metadata::<T>::insert(&asset_id, &metadata);
```
We could add a judegment before insert operation, if asset_id exist, throw AssetIdExist Error.

zqhxuyuan replied:
I'm not sure `try_mutate` or `insert` which one is best:
```
Metadata::<T>::try_mutate(
	&asset_id,
	|maybe_asset_metadatas| -> DispatchResult {
		ensure!(maybe_asset_metadatas.is_none(), Error::<T>::AssetIdExisted);

		*maybe_asset_metadatas = Some(metadata.clone());
		Ok(())
	},
)?;

Ok(())
```

sander2 replied:
I added the check in `register_asset` - I named the error `ConflictingAssetId`, to be consistent with the existing `ConflictingLocation` error.

I also updated the code to use try_mutate where possible. Instead of `AssetIdNotExisted` I'm reusing the existing `AssetNotFound`

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet_prelude::*;
+use frame_support::traits::EnsureOrigin;
+use frame_support::transactional;
```
use tight format: `use frame_support::{....}`, also other one in this file and other files needs to tidy-up.

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+	) -> Option<AssetMetadata<T::Balance, T::CustomMetadata>> {
+		let asset_id = MultiLocationLookup::<T>::get(location)?;
+		Metadata::<T>::get(asset_id)
+	}
```
add another two helper method: `get_assetid(multilocation)` and `get_multilocation(asset_id)` which is helpful to `AssetId` <-> `MultiLocation` conversion.

sander2 replied:
`get_assetid` was already available through the getter generated by the storage item. I implemented `get_multilocation` but named it `get_multilocation` because of [this comment](https://github.com/open-web3-stack/open-runtime-module-library/pull/728#discussion_r867281890).

zqhxuyuan commented on traits/src/asset_registry.rs:
```diff
@@ -0,0 +1,14 @@
+}
+
+pub trait AssetProcessor<AssetId, Metadata> {
+	fn process_asset(id: Option<AssetId>, asset_metadata: Metadata) -> Result<(AssetId, Metadata), DispatchError>;
```
rename this method to `pre_register(..)` and add another method named: `post_register()`
```rust
fn pre_register(id: Option<AssetId>, asset_metadata: Metadata) -> Result<(AssetId, Metadata), DispatchError>;
fn post_register(id: Option<AssetId>, asset_metadata: Metadata) -> Result<(AssetId, Metadata), DispatchError>;
```

As customized parachains may want to do additional behaviour after asset registed.

sander2 replied:
Done (except the type is `fn post_register(_id: AssetId, _asset_metadata: Metadata) -> Result<(), DispatchError>`).

Currently they are only called in `register_asset`, not in `update_asset`. Not sure if we should have hooks there as well..

zqhxuyuan replied:
`update_asset` don't need hook for now. but if someone have those customize, it can be add then.

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+		) -> DispatchResult {
+			let _ = T::AuthorityOrigin::ensure_origin(origin)?;
+
+			let (asset_id, metadata) = T::OnRegisterAsset::process_asset(asset_id, metadata)?;
```
I think it'll be good if move `process_asset` processing part into `insert_metdata()`. for parachains who want to use their own dispatch call and intergrate orml, they can use `orml_asset_registry::Pallet::insert_metadata()`

sander2 replied:
That's a bit problematic, since `insert_metadata` is also used from within the `update_asset` function. Should we also call that in `update_asset` and `set_location`?

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+
+			let (asset_id, metadata) = T::OnRegisterAsset::process_asset(asset_id, metadata)?;
+
+			Self::insert_metadata(&asset_id, &metadata)?;
```
maybe rename `insert_metadata` to `do_register_asset`, and rename `update_metadata` to `do_update_asset`?

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+			Ok(())
+		})?;
+
+		if let Some(ref location) = metadata.location {
+			Self::do_insert_location(asset_id.clone(), location.clone())?;
+		}
```
move `do_insert_location` to inner of `Metadata::<T>::try_mutate()`

zqhxuyuan commented on asset-registry/src/mock/para.rs:
```diff
@@ -0,0 +1,337 @@
+pub type AssetRegistryWeightTrader =
+	(AssetRegistryTrader<FixedRateAssetRegistryTrader<MyFixedConversionRateProvider>, ToTreasury>,);
+
+pub struct MyFixedConversionRateProvider;
+impl FixedConversionRateProvider for MyFixedConversionRateProvider {
+	fn get_fee_per_second(location: &MultiLocation) -> Option<u128> {
+		let metadata = AssetRegistry::fetch_metadata_by_location(location)?;
+		Some(metadata.additional.fee_per_second)
+	}
+}
```
Can this move to impl.rs, I think this is a common implementation.

shaunxw commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+			asset_id: Option<T::AssetId>,
+		) -> DispatchResult {
+			let _ = T::AuthorityOrigin::ensure_origin(origin)?;
```
```suggestion
			T::AuthorityOrigin::ensure_origin(origin)?;
```

shaunxw commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+			match maybe_metadata {
+				None => Err(Error::<T>::AssetNotFound.into()),
+				Some(ref old_metadata) => {
+					// remove the old location lookup if it exists
+					if let Some(ref location) = old_metadata.location {
+						let location: MultiLocation =
+							location.clone().try_into().map_err(|()| Error::<T>::BadVersion)?;
+						LocationToAssetId::<T>::remove(location);
+					}
+
+					// insert new location
+					if let Some(ref location) = metadata.location {
+						Self::do_insert_location(asset_id.clone(), location.clone())?;
+					}
```
Need to check if location changed before remove/insert.

shaunxw commented on asset-registry/src/tests.rs:
```diff
@@ -0,0 +1,377 @@
+#![cfg(test)]
```
Unit tests should cover all pallet errors, the following is not covered yet:
- AssetNotFound
- BadVersion
- InvalidAssetId

sander2 replied:
It's not so easy to test the BadVersion error though - that's the error you get when converting locations fails. I am not even sure it is possible.. It's also not being tested in the rest of the project

shaunxw reviewed (commented):
I suggest to change `update_asset` dispatchable call signature to

```rust
pub fn update_asset(
			origin: OriginFor<T>,
			asset_id: T::AssetId,
			decimals: Option<u32>,
			name: Option<Vec<u8>>,
			// ... rest fields wrapped in `Option`
		) -> DispatchResult
```

The impl checks every field and update it if `Some`. Note `location` needs to be `Option<Option<VersionedMultiLocation>>`.

The rationale behind it is that in most cases, only one field will be updated, but the current call requires to copy all existing unchanged ones, which is unnecessary and prune to mistakes. Additionally, `set_asset_location` call can be merged into this one.

sander2 commented on asset-registry/src/impls.rs:
```diff
@@ -0,0 +1,162 @@
+
+// Return Existential deposit of an asset. Implementing this trait allows the
+// pallet to be used in the tokens::ExistentialDeposits config item
+impl<T: Config> GetByKey<T::AssetId, T::Balance> for Pallet<T> {
```
Done, although it needs to be generic over the Config, so usage is now `AssetRegistryExistentialDeposits::<Runtime>::get(asset_id)`

sander2 commented on asset-registry/src/impls.rs:
```diff
@@ -0,0 +1,162 @@
+				let new_weight = existing_weight.saturating_add(weight);
+
+				if let Some(amount) = W::convert_weight_to_fee(location, new_weight) {
+					let fee_increase = amount.saturating_sub(existing_fee);
```
> I might miss something, but not sure why adding weight to existing to get new_weight first, then convert to fee and subtract. Simply converting weight to fee should have the same result no?

It's equivalent only when `convert_weight_to_fee` is a linear function, which it is not guaranteed to be. Though maybe I'm overcomplicating things - I don't expect to see non-linear functions being used in practice..

> And the embedded if/if let checks could be flattened for readability, such as:
> ```if let Some(amount) = W::convert_weight_to_fee(location, new_weight) {```
> -->
> ```let amount = W::convert_weight_to_fee(location, new_weight).ok_or(XcmError::TooExpensive)?;```

That would stop the iteration over the assets in `payment.fungible`. The same applies for the last refactoring you suggest

shaunxw replied:
I don't see non-linear functions being used either. AFAIK ppl are all using fixed rate conversions, specifically the fixed rate weight traders implemented in Polkadot repo.

IMO we can simply aim for the same impl to provide a fixed rate trader. It's easier to use and maintain.

sander2 commented:
I think I've addressed all comments - PR should be ready for final review pass & merging

shaunxw commented on asset-registry/README.md:
```diff
@@ -0,0 +1,5 @@
+
+## Overview
+
+This module 
```
Missed this one in my previous review. Could you please finished the readme?

shaunxw commented on asset-registry/Cargo.toml:
```diff
@@ -0,0 +1,68 @@
+scale-info = { version = "2.0.1", default-features = false, features = ["derive"] }
+serde = { version = "1.0.136", optional = true }
+codec = { package = "parity-scale-codec", version = "3.0.0", default-features = false, features = ["max-encoded-len"] }
+sp-runtime = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.18", default-features = false }
```
Please also update deps to `v0.9.20`.

shaunxw reviewed (commented):
LGTM. A few more trivial issues and I think we are good to go.

sander2 commented:
Fixed the last comments and rebased (and squashed) the branch.

shaunxw commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,312 @@
+			}
+		})?;
+
+		// Self::deposit_event(Event::<T>::UpdatedAsset { asset_id, metadata });
```
Will need to deposit event, probably in `Some(ref mut metadata) => {` branch.

sander2 replied:
Oops, fixed

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,316 @@
+
+	// This is useful in tests but it might also come in useful to users of this
+	// pallet.
+	/// Like do_register_asset, but without calling pre_register and
```
use `///` above

sander2 commented on asset-registry/src/impls.rs:
```diff
@@ -0,0 +1,169 @@
+impl<P: FixedConversionRateProvider> WeightToFeeConverter for FixedRateAssetRegistryTrader<P> {
+	fn convert_weight_to_fee(location: &MultiLocation, weight: Weight) -> Option<u128> {
+		let fee_per_second = P::get_fee_per_second(location)?;
+		let amount = fee_per_second.saturating_mul(weight as u128) / (WEIGHT_PER_SECOND as u128);
```
I didn't think that was necessary since WEIGHT_PER_SECOND is a constant, but I made the change you requested regardless

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,316 @@
+			name: Option<Vec<u8>>,
+			symbol: Option<Vec<u8>>,
+			existential_deposit: Option<T::Balance>,
+			location: Option<Option<VersionedMultiLocation>>,
```
we can simplify this parameter as `Option<VersionedMultiLocation>`.

sander2 replied:
We can't - the outer Option determines whether or not to update that parameter, whereas the inner one is there because the user might want to unset the location, i.e. to make `metadata.location = None`

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,316 @@
+	) -> DispatchResult {
+		Metadata::<T>::try_mutate(&asset_id, |maybe_metadata| -> DispatchResult {
+			match maybe_metadata {
+				None => Err(Error::<T>::AssetNotFound.into()),
```
replace with below, and remove match ase. also replace `(*metadata)` with `metadata`
```rust
let metadata = maybe_metadata.as_mut().ok_or(Error::<T>::AssetNotFound)?;
```

zqhxuyuan commented on asset-registry/src/lib.rs:
```diff
@@ -0,0 +1,316 @@
+					}
+
+					if let Some(location) = location {
+						// Update LocationToAssetId only if location changed
+						if location != metadata.location {
+							// remove the old location lookup if it exists
+							if let Some(ref location) = metadata.location {
+								let location: MultiLocation =
+									location.clone().try_into().map_err(|()| Error::<T>::BadVersion)?;
+								LocationToAssetId::<T>::remove(location);
+							}
+
+							// insert new location
+							if let Some(ref new_location) = location {
+								Self::do_insert_location(asset_id.clone(), new_location.clone())?;
+							}
+						}
```
would be better to use `LocationToAssetId::try_mutate(location, |maybe_asset_id| {....})`

sander2 replied:
We are dealing with 2 different keys though - we have `old_location` and `new_location`, so it can't be combined into a single `try_mutate`. Setting `new_location` does use `try_mutate` inside `do_insert_location`, but for `old_location` the remove is more straightforward
